### PR TITLE
Get rid of `Buffer` and use an associated type in the `RequestConnection` trait

### PR DESCRIPTION
**Pull request by eduardosm**

### CODE REVIEW

**Comment by eduardosm** [comment]
Regarding `GenericEvent` and `GenericErrors`, I had to remove the `TryFrom` and `Into` implementations for the same reasons as in https://github.com/psychon/x11rb/issues/251#issuecomment-596100612.

File: src/cookie.rs [hunk 1]
```diff
@@ -292,7 +292,7 @@ where
         let reply = ListFontsWithInfoReply::try_from(reply.as_ref()).map_err(ReplyError::from);
         if reply.is_ok() {
             // Is this an indicator that no more replies follow?
-            if !reply.as_ref().unwrap().name.is_empty() {
+            if !reply.as_ref().ok().unwrap().name.is_empty() {
```
Comment (psychon) [inline src/cookie.rs]: Huh? This is `Result::ok()`, I think. Seems unnecessary to me, or am I missing something?

File: src/utils.rs [hunk 2]
```diff
@@ -20,7 +18,6 @@ mod unsafe_code {
     /// Wrapper around a slice that was allocated in C code.
     ///
     /// `CSlice` is only available when the `allow-unsafe-code` feature is enabled.
-    #[derive(Debug)]
```
Comment (psychon) [inline src/utils.rs]: Hm?

Comment (psychon) [reply src/utils.rs]: Well, okay, I guess the `Debug` implementation for `NonNull` is not really useful, so this is actually a good idea, I guess...

File: src/cookie.rs (hunk 1 above)
Comment (eduardosm) [reply src/cookie.rs]: Indeed, it is not needed now. I added it was at some point, but I'll remove it.

File: src/utils.rs (hunk 2 above)
Comment (psychon) [reply src/utils.rs]: Oh, ups. A hand written implementation is added below. I should start using GitHub PR reviews...

Comment (eduardosm) [reply src/utils.rs]: Exactly, the derived implementation was just printing the pointer address.

**Comment by psychon** [comment]
I am not too happy about all the errors and events now needing a type parameter, but... meh, whatever. It's worth it.

File: src/cookie.rs (hunk 1 above)
Comment (eduardosm) [reply src/cookie.rs]: Changed

**Comment by eduardosm** [comment]
> I am not too happy about all the errors and events now needing a type parameter, but... meh, whatever. It's worth it.

I'm afraid its the only way. But I also think it is worth.

**Pull request merged**